sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/feedranges/FeedRangeEpkImpl.java

```diff
@@ -3,125 +3,253 @@
+
+                final DocumentCollection collection = documentCollectionResourceResponse.v;
+                if (collection == null) {
+                    throw new IllegalStateException("Collection cannot be null");
                 }
 
-                return Mono.just((UnmodifiableList<String>)UnmodifiableList.unmodifiableList(rangeList));
+                final String containerRid = collection.getResourceId();
+
+                return routingMapProvider
+                    .tryGetOverlappingRangesAsync(
+                        metadataDiagnosticsCtx,
+                        containerRid,
+                        this.range,
+                        false,
+                        null)
+                    .flatMap(pkRangeHolder -> {
+                        final ArrayList<String> rangeList = new ArrayList<>();
+
+                        if (pkRangeHolder != null) {
+                            final List<PartitionKeyRange> pkRanges = pkRangeHolder.v;
+                            for (final PartitionKeyRange pkRange : pkRanges) {
+                                rangeList.add(pkRange.getId());
+                            }
+                        }
+
+                        return Mono.just(UnmodifiableList.unmodifiableList(rangeList));
+                    });
             });
     }
 
     @Override
-    public String toString() {
-        return this.range.toString();
+    public Mono<RxDocumentServiceRequest> populateFeedRangeFilteringHeaders(
+        IRoutingMapProvider routingMapProvider,
+        RxDocumentServiceRequest request,
+        Mono<Utils.ValueHolder<DocumentCollection>> collectionResolutionMono) {
+
+        checkNotNull(
+            routingMapProvider,
+            "Argument 'routingMapProvider' must not be null");
+        checkNotNull(
+            request,
+            "Argument 'request' must not be null");
+        checkNotNull(
+            collectionResolutionMono,
+            "Argument 'collectionResolutionMono' must not be null");
+
+        MetadataDiagnosticsContext metadataDiagnosticsCtx =
+            BridgeInternal.getMetaDataDiagnosticContext(request.requestContext.cosmosDiagnostics);
+
+        return collectionResolutionMono
+            .flatMap(documentCollectionResourceResponse -> {
+
+                final DocumentCollection collection = documentCollectionResourceResponse.v;
+                if (collection == null) {
+                    throw new IllegalStateException("Collection cannot be null");
+                }
+
+                final String containerRid = collection.getResourceId();
+
+                return routingMapProvider
+                    .tryGetOverlappingRangesAsync(
+                        metadataDiagnosticsCtx,
+                        containerRid,
+                        this.range,
+                        false,
+                        null)
+                    .flatMap(pkRangeHolder -> {
+                        if (pkRangeHolder == null) {
```

Comment:
we will never hit here. as explained above null is not a permissible value for reactive-stream

Reply:
Fixed in the next iteration